### PR TITLE
Accelerated DAG: Support dynamic resizing of shared memory channels

### PR DESCRIPTION
**Pull request by jackhumphries**
## Why are these changes needed?

In an accelerated DAG, all channels have their backing store specific upfront. However, this size may not be large enough for all objects that are communicated via channels, particularly since channels are reused across multiple DAG invocations.

This PR adds support for increasing the size of a shared memory channel backing store. Once the backing store size is increased, the size is maintained (i.e., it is not decreased) until it needs to be increased again.

## Related issue number

Addresses one issue noted in #45597

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/experimental/channel/shared_memory_channel.py [hunk 1]
```diff
@@ -323,6 +328,42 @@ def write(self, value: Any):
...
+        size = serialized_value.total_bytes + len(serialized_value.metadata)
+        if size > self._typ.buffer_size_bytes:
+            # Now make the channel backing store larger.
+            self._typ.buffer_size_bytes = size
+            # TODO(jhumphri): I assume the current writer ref object is automatically
+            # freed when the reference to it is destroyed in the line below?
+            prev_writer_ref = self._writer_ref
+            self._writer_ref = _create_channel_ref(self, self._typ.buffer_size_bytes)
+            # We need to register the new writer_ref.
+            self.ensure_registered_as_writer(force=True)
```
Comment (stephanie-wang) [inline python/ray/experimental/channel/shared_memory_channel.py:343]: Nit: resetting `self._writer_registered = False` seems a little cleaner (so future callers don't need to figure our if they need to set `force` or not).

File: python/ray/experimental/channel/shared_memory_channel.py [hunk 2]
```diff
@@ -323,6 +328,42 @@ def write(self, value: Any):
             )
             raise TypeError(msg) from e
 
+        # serialized_value.total_bytes *only* includes the size of the data. It does not
+        # include the size of the metadata, so we must account for the size of the
+        # metadata explicitly.
+        size = serialized_value.total_bytes + len(serialized_value.metadata)
+        if size > self._typ.buffer_size_bytes:
+            # Now make the channel backing store larger.
```
Comment (stephanie-wang) [inline python/ray/experimental/channel/shared_memory_channel.py:336]: Probably would be best to separate this out into a helper method.

File: python/ray/experimental/channel/shared_memory_channel.py [hunk 3]
```diff
@@ -323,6 +328,42 @@ def write(self, value: Any):
...
+            self._typ.buffer_size_bytes = size
+            # TODO(jhumphri): I assume the current writer ref object is automatically
+            # freed when the reference to it is destroyed in the line below?
+            prev_writer_ref = self._writer_ref
+            self._writer_ref = _create_channel_ref(self, self._typ.buffer_size_bytes)
```
Comment (stephanie-wang) [inline python/ray/experimental/channel/shared_memory_channel.py:341]: Suggest separating out object allocation into a helper method too, so that it can be shared between this and the constructor.

File: python/ray/experimental/channel/shared_memory_channel.py (hunk 1 above)
Comment (jackhumphries) [reply python/ray/experimental/channel/shared_memory_channel.py:343]: Done.

File: python/ray/experimental/channel/shared_memory_channel.py (hunk 2 above)
Comment (jackhumphries) [reply python/ray/experimental/channel/shared_memory_channel.py:322]: Done.

File: python/ray/experimental/channel/shared_memory_channel.py (hunk 3 above)
Comment (jackhumphries) [reply python/ray/experimental/channel/shared_memory_channel.py:362]: Could you please elaborate on how to do this? It's not clear to me how to decompose this beyond what we have in _create_channel_ref(). Thanks

**Comment by jackhumphries** [comment]
This is ready for another pass.

File: python/ray/experimental/channel/shared_memory_channel.py (hunk 3 above)
Comment (stephanie-wang) [reply python/ray/experimental/channel/shared_memory_channel.py:362]: Ah it seemed to me that this logic [here](https://github.com/ray-project/ray/blob/08b4b49de8e1ca8466abad91ad85953f3042e224/python/ray/experimental/channel/shared_memory_channel.py#L219-L239) and [here](https://github.com/ray-project/ray/blob/08b4b49de8e1ca8466abad91ad85953f3042e224/python/ray/experimental/channel/shared_memory_channel.py#L362-L374) was basically the same. I haven't looked super carefully though, so maybe not worth factoring out.

Comment (jackhumphries) [reply python/ray/experimental/channel/shared_memory_channel.py:365]: Done.

File: python/ray/experimental/channel/shared_memory_channel.py [hunk 4]
```diff
@@ -259,6 +264,15 @@ def __init__(
             self.ensure_registered_as_writer()
             assert self._reader_ref is not None
 
+    def create_reader_ref(self, readers, buffer_size_bytes):
```
Comment (stephanie-wang) [inline python/ray/experimental/channel/shared_memory_channel.py:267]:
```suggestion
    def _create_reader_ref(self, readers, buffer_size_bytes):
```

(Use underscore for internal methods)

Can you also add type hints?

File: python/ray/experimental/channel/shared_memory_channel.py [hunk 5]
```diff
@@ -337,6 +351,37 @@ def __reduce__(self):
...
+    def resize_channel_if_needed(self, serialized_value):
+        # serialized_value.total_bytes *only* includes the size of the data. It does not
+        # include the size of the metadata, so we must account for the size of the
+        # metadata explicitly.
+        size = serialized_value.total_bytes + len(serialized_value.metadata)
+        if size > self._typ.buffer_size_bytes:
+            # Now make the channel backing store larger.
+            self._typ.buffer_size_bytes = size
+            # TODO(jhumphri): I assume the current writer ref object is automatically
+            # freed when the reference to it is destroyed in the line below?
```
Comment (stephanie-wang) [inline python/ray/experimental/channel/shared_memory_channel.py:363]: I don't think it will get automatically freed. Can you change this to a TODO to handle garbage collection?

File: python/ray/tests/test_channel.py [hunk 6]
```diff
@@ -140,11 +140,137 @@ def _test(val):
     _test(1000)
     _test(np.random.rand(10))
 
-    # Cannot put a serialized value larger than the allocated buffer.
-    with pytest.raises(ValueError):
-        _test(np.random.rand(100))
 
-    _test(np.random.rand(1))
+def test_multiple_channels_different_nodes(ray_start_cluster):
```
Comment (stephanie-wang) [inline python/ray/tests/test_channel.py:144]: Can you add a docstring to explain what the test does?

File: python/ray/experimental/channel/shared_memory_channel.py [hunk 7]
```diff
@@ -80,6 +83,11 @@ def _create_channel_ref(
     return object_ref
 
 
+class ResizeChannel:
```
Comment (stephanie-wang) [inline python/ray/experimental/channel/shared_memory_channel.py:86]: Add docstring.

Also, we should rename to `_ResizeChannel` since this class is only used internally.

File: src/ray/core_worker/experimental_mutable_object_provider.cc [hunk 8]
```diff
@@ -74,20 +84,14 @@ void MutableObjectProvider::HandleRegisterMutableObject(
...
-    LocalReaderInfo info;
-    info.num_readers = num_readers;
-    info.local_object_id = reader_object_id;
-    bool success =
-        remote_writer_object_to_local_reader_.insert({writer_object_id, info}).second;
-    RAY_CHECK(success);
-
-    RegisterReaderChannel(reader_object_id);
-  }
+  LocalReaderInfo info;
```
Comment (stephanie-wang) [inline src/ray/core_worker/experimental_mutable_object_provider.cc:87]: Why the change from the previous code here?

File: src/ray/core_worker/experimental_mutable_object_provider.h [hunk 9]
```diff
@@ -165,18 +162,19 @@ class MutableObjectProvider {
...
-  // the changes to remote nodes via the network.
-  instrumented_io_context io_service_;
-  // Manages RPCs for inter-node communication of mutable objects.
-  boost::asio::io_service::work io_work_;
-  // Manages outgoing RPCs that send mutable object changes to remote nodes.
-  std::unique_ptr<rpc::ClientCallManager> client_call_manager_;
-  // Thread that waits for local mutable object changes and then sends the changes to
-  // remote nodes via the network.
-  std::thread io_thread_;
+  // Manage RPCs for inter-node communication of mutable objects.
```
Comment (stephanie-wang) [inline src/ray/core_worker/experimental_mutable_object_provider.h:168]: Can you add a comment explaining why we need to store a vector of threads instead of just one?

Also, I don't this comment is accurate; doesn't `work` should keep the event loop alive, not "manage RPCs"?

File: python/ray/experimental/channel/shared_memory_channel.py [hunk 10]
```diff
@@ -337,6 +351,37 @@ def __reduce__(self):
             self._reader_ref,
         )
 
+    def resize_channel_if_needed(self, serialized_value):
```
Comment (stephanie-wang) [inline python/ray/experimental/channel/shared_memory_channel.py:354]:
```suggestion
    def _resize_channel_if_needed(self, serialized_value):
```
Add type hint.

File: python/ray/experimental/channel/shared_memory_channel.py [hunk 11]
```diff
@@ -337,6 +351,37 @@ def __reduce__(self):
...
+            # TODO(jhumphri): I assume the current writer ref object is automatically
+            # freed when the reference to it is destroyed in the line below?
+            prev_writer_ref = self._writer_ref
+            self._writer_ref = _create_channel_ref(self, self._typ.buffer_size_bytes)
+
+            self.create_reader_ref(self._readers, self._typ.buffer_size_bytes)
+
+            # We need to register the new writer_ref.
+            self._writer_registered = False
+            self.ensure_registered_as_writer()
```
Comment (stephanie-wang) [inline python/ray/experimental/channel/shared_memory_channel.py:371]: Can we also move this logic into `create_reader_ref`?

Comment (jackhumphries) [reply python/ray/experimental/channel/shared_memory_channel.py]: Done.

File: python/ray/experimental/channel/shared_memory_channel.py (hunk 4 above)
Comment (jackhumphries) [reply python/ray/experimental/channel/shared_memory_channel.py]: Done.

File: python/ray/experimental/channel/shared_memory_channel.py (hunk 10 above)
Comment (jackhumphries) [reply python/ray/experimental/channel/shared_memory_channel.py]: Done.

File: python/ray/experimental/channel/shared_memory_channel.py (hunk 5 above)
Comment (jackhumphries) [reply python/ray/experimental/channel/shared_memory_channel.py]: Done.

File: src/ray/core_worker/experimental_mutable_object_provider.cc (hunk 8 above)
Comment (jackhumphries) [reply src/ray/core_worker/experimental_mutable_object_provider.cc:87]: Happy to change it back, but it seemed cleaner to make this idempotent rather than change the functionality based on whether the mutable object has already been scheduled.

File: src/ray/core_worker/experimental_mutable_object_provider.h (hunk 9 above)
Comment (jackhumphries) [reply src/ray/core_worker/experimental_mutable_object_provider.h]: Good catch! Done.

File: python/ray/experimental/channel/shared_memory_channel.py (hunk 7 above)
Comment (jackhumphries) [reply python/ray/experimental/channel/shared_memory_channel.py]: Done.

File: python/ray/tests/test_channel.py (hunk 6 above)
Comment (jackhumphries) [reply python/ray/tests/test_channel.py:149]: Done. Do you want me to add docstrings for all of the tests?

**Review by kevin85421 — COMMENTED** [review]
The Python changes overall look good to me. I am still trying to understand the C++ changes.

File: python/ray/experimental/channel/shared_memory_channel.py [hunk 12]
```diff
@@ -185,6 +193,9 @@ def __init__(
                 "bytes or a SharedMemoryType"
             )
 
+        if typ.buffer_size_bytes < MIN_BUFFER_SIZE:
+            raise ValueError("typ.buffer_size_bytes must be at least MIN_BUFFER_SIZE")
```
[the remark this answers is not in the document]
Comment (jackhumphries) [reply python/ray/experimental/channel/shared_memory_channel.py]: Done.

File: python/ray/tests/test_channel.py (hunk 6 above)
Comment (stephanie-wang) [reply python/ray/tests/test_channel.py:149]: That would be great! I don't always do it if the test name is straightforward enough, but it's a good practice.

File: src/ray/core_worker/experimental_mutable_object_provider.cc (hunk 8 above)
Comment (stephanie-wang) [reply src/ray/core_worker/experimental_mutable_object_provider.cc:87]: No this is alright; that was just me skimming quickly and not being to immediately tell what the difference was.

File: python/ray/experimental/channel/shared_memory_channel.py [hunk 13]
```diff
@@ -185,6 +199,22 @@ def __init__(
                 "bytes or a SharedMemoryType"
             )
 
+        if typ.buffer_size_bytes < MIN_BUFFER_SIZE:
+            raise ValueError(
+                "typ.buffer_size_bytes must be at least MIN_BUFFER_SIZE (%d bytes)"
```
Comment (stephanie-wang) [inline python/ray/experimental/channel/shared_memory_channel.py:204]:
```suggestion
                f"typ.buffer_size_bytes must be at least MIN_BUFFER_SIZE ({MIN_BUFFER_SIZE} bytes)"
```

File: python/ray/tests/test_channel.py (hunk 6 above)
Comment (stephanie-wang) [reply python/ray/tests/test_channel.py:149]: Also, typically we put the docstring on the line after `def` for python.

**Review by stephanie-wang — APPROVED** [review]
Nice, very clean :)

File: src/ray/core_worker/experimental_mutable_object_provider.cc [hunk 14]
```diff
@@ -177,14 +183,19 @@ void MutableObjectProvider::PollWriterClosure(
       object->GetData()->Size(),
       object->GetMetadata()->Size(),
```
Comment (kevin85421) [inline src/ray/core_worker/experimental_mutable_object_provider.cc:184]: > `raylet_client_factory_`: Creates a function for each object. This function waits for changes on the object and then sends those changes to a remote node via RPC.

The above comment is from `experimental_mutable_object_provider.h`. Can you explain how does `reader->PushMutableObject` "sends those changes to a remote node via RPC"? Thanks!

**Review by kevin85421 — COMMENTED** [review]
I am still trying to understand the C++ parts.

File: src/ray/core_worker/experimental_mutable_object_provider.cc [hunk 15]
```diff
@@ -49,14 +49,24 @@ void MutableObjectProvider::RegisterWriterChannel(const ObjectID &object_id,
   if (node_id) {
     // Start a thread that repeatedly listens for values on this object and then sends
     // them via RPC to the remote reader.
+
+    io_services_.push_back(std::make_unique<instrumented_io_context>());
+    instrumented_io_context &io_service = *io_services_.back();
+    io_works_.push_back(std::make_unique<boost::asio::io_service::work>(io_service));
```
[the remark this answers is not in the document]
Comment (kevin85421) [reply src/ray/core_worker/experimental_mutable_object_provider.cc:55]: `boost::asio::io_service::work` -> `boost::asio::io_context::work`

File: src/ray/core_worker/experimental_mutable_object_provider.cc (hunk 14 above)
Comment (jackhumphries) [reply src/ray/core_worker/experimental_mutable_object_provider.cc:184]: I fixed the comment.

File: src/ray/core_worker/experimental_mutable_object_provider.cc (hunk 15 above)
[the remark this answers is not in the document]
Comment (jackhumphries) [reply src/ray/core_worker/experimental_mutable_object_provider.cc]: I ended up changing this to call `work.reset()` instead. The mutable object provider experienced a segfault in its destructor when I called `io_context.run()`. I think the work_guard triggers this if it is not reset but its corresponding thread is stopped.
